### PR TITLE
#2132: Relative URIs for @oData.context are not supported

### LINKED ISSUE #2132: Relative URIs for @oData.context are not supported
The lib expects an absolute URI in the @oData.context field of an entity. If the server uses relative URIs, parsing of the response fails.

### Assemblies affected

OData .Net lib 7.9.0 and earlier

### Reproduce steps

Use an oData service which uses relative URIs in the @oData.context field (for example SAP oData V4 service; mine is password protected so I can't share it).

### Expected result

The entities should be returned.

### Actual result

Parsing of response fails with an InvalidOperationException: `The response payload is a not a valid response payload. Please make sure that the top level element is a valid Atom or JSON element or belongs to 'http://docs.oasis-open.org/odata/ns/data' namespace.`

### Additional detail

An ODataException is thrown at ODataJsonLightContextUriParser.cs line 86:

```
Uri contextUri;
if (!Uri.TryCreate(contextUriFromPayload, UriKind.Absolute, out contextUri))
{
        throw new ODataException(ODataErrorStrings.ODataJsonLightContextUriParser_TopLevelContextUrlShouldBeAbsolute(contextUriFromPayload));
}
```

The ODataException is eaten at ODataJsonLightPayloadKindDetectionDeserializer on line 57 and an InvalidOperationException is thrown at ODataMaterializer on line 359.

### PR DESCRIPTION
**Pull request by karimabd84**
### Issues

*This pull request fixes #2132 .*
The issue was created from my college https://github.com/chris472

### Description

*The lib expects an absolute URI in the @oData.context field of an entity. If the server uses relative URIs, parsing of the response fails.*

### CODE REVIEW

File: src/Microsoft.OData.Core/ODataDeserializer.cs [hunk 1]
```diff
@@ -68,6 +68,17 @@ internal IEdmModel Model
...
         }
 
+        /// <summary>
+        /// Get the absolute uri
+        /// </summary>
+        internal string AbsoluteUri
+        {
+            get
+            {
+                return this.inputContext?.MessageReaderSettings?.BaseUri?.AbsoluteUri;
```
Comment (corranrogue9) [inline src/Microsoft.OData.Core/ODataDeserializer.cs:78]: Given the other properties, I think it is safe to assume `inputContext` is not `null`

**Review by corranrogue9 — APPROVED** [review]

File: src/Microsoft.OData.Core/JsonLight/ODataJsonLightDeserializer.cs [hunk 2]
```diff
@@ -240,6 +241,17 @@ internal static bool TryParsePropertyAnnotation(string propertyAnnotationName, o
                 isReadingNestedPayload,
                 allowEmptyPayload);
 
+            //https://github.com/OData/odata.net/issues/2132
```
Comment (KenitoInc) [inline src/Microsoft.OData.Core/JsonLight/ODataJsonLightDeserializer.cs:244]: The first 2 comments are not necessary

**Review by KenitoInc — CHANGES_REQUESTED** [review]
Add tests

**Comment by tywunon** [comment]
Any Idea when this is going to be released?

File: src/Microsoft.OData.Core/JsonLight/ODataJsonLightDeserializer.cs [hunk 3]
```diff
@@ -240,6 +241,17 @@ internal static bool TryParsePropertyAnnotation(string propertyAnnotationName, o
                 isReadingNestedPayload,
                 allowEmptyPayload);
 
+            //https://github.com/OData/odata.net/issues/2132
+            //Relative URIs for @oData.context are not supported #2132
+            //Concate the absoluteUri with the contextUri, when the contextUri is arelativeUri.
```
Comment (tywunon) [inline src/Microsoft.OData.Core/JsonLight/ODataJsonLightDeserializer.cs:246]: Typo? arelativeUri -> a relativeUri

**Comment by tywunon** [comment]
Hey Guys,

any update on this?

I am waiting for this, because a client is not able to switch to absolute paths, and I would like to use this Library for my project. Without this fix, i cannot use it. Can you please review this and release it?

Thanks.

**Comment by KenitoInc** [comment]
> Hey Guys,
> 
> any update on this?
> 
> I am waiting for this, because a client is not able to switch to absolute paths, and I would like to use this Library for my project. Without this fix, i cannot use it. Can you please review this and release it?
> 
> Thanks.

I am waiting for @karimabd84 to add tests

**Comment by karimabd84** [comment]
@KenitoInc Which project can use to add my tests?

**Comment by KenitoInc** [comment]
> @KenitoInc Which project can use to add my tests?

OData.sln has tests

**Comment by tywunon** [comment]
Hey, sorry for bugging you all about this.
But if this can't get through by Monday, I would have to start looking for alternatives.
If you can tell me this will not be the case anyway, I could start looking.

Thanks, and sorry.

**Comment by KenitoInc** [comment]
> Hey, sorry for bugging you all about this. But if this can't get through by Monday, I would have to start looking for alternatives. If you can tell me this will not be the case anyway, I could start looking.
> 
> Thanks, and sorry.

I would to merge and release this PR, unfortunately I can't do that without tests. cc @karimabd84 
Let me see if we can get a resource to do that

**Comment by tywunon** [comment]
Sorry for pushing this. I would just make my life way easier.

**Comment by tywunon** [comment]
> 

And I totally get why you just can't merge this.

File: src/Microsoft.OData.Core/ODataDeserializer.cs (hunk 1 above)
Comment (xuzhg) [reply src/Microsoft.OData.Core/ODataDeserializer.cs]: inputContext is from te ctor and there's a null check but using Debug.Assert.

File: src/Microsoft.OData.Core/JsonLight/ODataJsonLightDeserializer.cs [hunk 4]
```diff
@@ -240,6 +241,20 @@ internal static bool TryParsePropertyAnnotation(string propertyAnnotationName, o
...
 
+            //Concatenate the metadata uri with the contextUri, when the contextUri is a relativeUri. 
+            Uri contextUri;
+            if (!string.IsNullOrEmpty(this.BaseUri) && 
+                !string.IsNullOrEmpty(contextUriAnnotationValue) &&                 
+                !Uri.TryCreate(contextUriAnnotationValue, UriKind.Absolute, out contextUri))
+            {
+                // If the Base uri string is http://odata.org/test
+                // The MetadataDocumentUri will be http://odata.org/test/$metadata
+                // If the contextUriAnnotation valie is Customers(1)/Name
```
Comment (xuzhg) [inline src/Microsoft.OData.Core/JsonLight/ODataJsonLightDeserializer.cs:252]: typo

File: src/Microsoft.OData.Core/ODataDeserializer.cs [hunk 5]
```diff
@@ -68,6 +68,17 @@ internal IEdmModel Model
             }
         }
 
+        /// <summary>
+        /// Get the base uri.
+        /// </summary>
+        internal string BaseUri
```
Comment (xuzhg) [inline src/Microsoft.OData.Core/ODataDeserializer.cs:74]: >string [](http://example.com/codeflow?start=17&length=6)

why just return "Uri" type?

**Review by xuzhg — APPROVED** [review]
:shipit:

**Review by KenitoInc — APPROVED** [review]

**Pull request merged**